src/common/helpers/application-lock.js

```diff
@@ -0,0 +1,188 @@
+import { config } from '../../config.js'
+
+export const LOCK_TTL_MS = config.get('applicationLock.ttlMs')
+
+/**
+ * Builds a deterministic lock identifier for an application.
+ *
+ * @param {string} grantCode - Grant scheme code
+ * @param {number} grantVersion - Grant scheme version
+ * @param {string} sbi - Single Business Identifier
+ * @returns {string} Application lock identifier
+ */
+export function getApplicationLockId(grantCode, grantVersion, sbi) {
+  return `app-lock:${grantCode}:${grantVersion}:${sbi}`
+}
+
+/**
+ * Attempts to acquire an existing lock if:
+ *  - the lock has expired, or
+ *  - the lock is already owned by the same user (re-entrant access).
+ *
+ * This operation is atomic and does not create new lock documents.
+ *
+ * @param {import('mongodb').Collection} collection - MongoDB lock collection
+ * @param {Object} params
+ * @param {string} params.grantCode
+ * @param {number} params.grantVersion
+ * @param {string} params.sbi
+ * @param {string} params.ownerId - DefraID user ID
+ * @param {Object} timing
+ * @param {Date} timing.now
+ * @param {Date} timing.expiresAt
+ * @returns {Promise<Object|null>} The updated lock document, or null if not acquirable
+ */
+async function tryAcquireExistingLock(collection, { grantCode, grantVersion, sbi, ownerId }, { now, expiresAt }) {
+  const result = await collection.findOneAndUpdate(
+    {
+      grantCode,
+      grantVersion,
+      sbi,
+      $or: [{ expiresAt: { $lte: now } }, { ownerId }]
+    },
+    {
+      $set: {
+        ownerId,
+        lockedAt: now,
+        expiresAt
+      }
+    },
+    { returnDocument: 'after' }
+  )
+
+  return result?.value ?? null
```

Comment:
Should not be .value - https://www.mongodb.com/community/forums/t/findoneandupdate-return-value-property-no-longer-exists/257780

Reply:
OK.